### PR TITLE
Shinwool/upstream from bionemo utils scheduler

### PR DESCRIPTION
**Pull request by citygyraffe**
# What does this PR do ?

Adds a LR scheduler policy (LinearWarmupHoldDecayPolicy) that has following properties:
- Phase 1: linear warmup from 0 to the optimizer's learning rate
- Phase 2: constant phase
- Phase 3: linear decay to 'min_lr'
- Phase 4: constant phase at 'min_lr'
A visual representation of above is included in code.

**Collection**: N/A

# Changelog 
- Added class for LR scheduler policy (LinearWarmupHoldDecayPolicy)
- Added unit test for above

# Usage

# Before your PR is "Ready for review"
**Pre checks**:
- [v] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [v] Did you write any new necessary tests?
- [v] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [v] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/core/optim/lr_scheduler.py [hunk 1]
```diff
@@ -640,6 +640,119 @@ def _get_lr(self, step):
...
+        min_lr=0.0,
+        last_epoch=-1,
+    ):
+        assert not (
+            warmup_steps is not None and warmup_ratio is not None
+        ), "Either use particular number of step or ratio"
+        assert not (
+            constant_steps is not None and constant_ratio is not None
+        ), "Either use constant_steps or constant_ratio"
+        assert warmup_ratio is None or max_steps is not None, "If there is a " "ratio, there should be a total steps"
```
Comment (ericharper) [inline nemo/core/optim/lr_scheduler.py:696]: Could you use f-strings to display the values of the offending variables in the error message?

**Comment by citygyraffe** [comment]
keep-alive

**Comment by ericharper** [comment]
jenkins

**Review by ericharper — CHANGES_REQUESTED** [review]
Just one small comment. Please use f-strings.

File: nemo/core/optim/lr_scheduler.py [hunk 2]
```diff
@@ -640,6 +640,119 @@ def _get_lr(self, step):
         return new_lrs
```
Comment (titu1994) [inline nemo/core/optim/lr_scheduler.py:642]: Name of such class is by conversion "Annealing' at the end, not Policy. Though it's fine since it's just convention
